Tasks/Common/webdeployment-common-v2/ziputility.ts

```diff
@@ -2,13 +2,87 @@ import tl = require('azure-pipelines-task-lib/task');
 import path = require('path');
 import Q = require('q');
 import fs = require('fs');
-import { ToolRunner } from 'azure-pipelines-task-lib/toolrunner';
+import tr = require('azure-pipelines-task-lib/toolrunner');
 
 var DecompressZip = require('decompress-zip');
 var archiver = require('archiver');
 
 const deleteDir = (path: string) => tl.exist(path) && tl.rmRF(path);
 
+const extractWindowsZip = async (fromFile: string, toDir: string, usePowerShell?: boolean) => {
+    if (usePowerShell) {
+        await extractUsingPowerShell(fromFile, toDir);
+    }
+    else {
+        await extractUsing7zip(fromFile, toDir);
+    }
+}
+
+const extractUsingPowerShell = async (fromFile: string, toDir: string) => {
+    tl.debug(`Using PowerShell for extracting zip ${fromFile}`);
+    let command = `Expand-Archive -Path "${fromFile}" -DestinationPath "${toDir}"`;
+    tl.debug(`Command to execute: '${command}'`)
+    let powershellPath: string = ''
+    let packageSize: number = 0
+
+    try {
+        let packageStats: fs.Stats = fs.statSync(fromFile)
+        // size in mb
+        packageSize = Math.floor(packageStats.size / 1000000)
+    }
+    catch (error) {
+        tl.debug("Error occurred while trying to calculate package size in MB.")
+        tl.debug(error)
+        packageSize = -1
+    }
+
+    tl.debug(`Package Size = '${packageSize}' MB`)
+
+    try {
+        powershellPath = tl.which('pwsh', true)
+    }
+    catch (error) {
+        tl.debug(`Tool 'pwsh' not found. Error: ${error}`)
+        tl.debug("PowerShell core is not available on agent machine. Falling back to using Windows PowerShell.")
```

Comment:
Please add a code comment with the reason why we prefer pwsh and then fallback to powershell.